internal/dispatch/singleflight/singleflight.go

```diff
@@ -3,8 +3,11 @@ package singleflight
 import (
 	"context"
 	"encoding/hex"
+	"errors"
+	"fmt"
 	"strconv"
 
+	"github.com/bits-and-blooms/bloom/v3"
```

Comment:
Is this library better than https://github.com/tylertreat/BoomFilters which has been around for years and has many different types of probabilistic data structures?

Reply:
@jzelinskie I looked into that library, it's nice it has a lot of more data structures than just the classic bloom filter. From the API they are basically on par, but bits-and-blooms provides convenience AddString methods that spares you the type-casting.

From the performance point of view, bits-and-blooms seems significantly better, both time and space wise. I created 2 benchmarks that take 50 random strings, adds them, tests them and then serializes the filter, both on binary and in Gob serialization. This is what I got using `benchstat`:

```
benchstat bits-and-blooms.txt boomfilters.txt         
goos: darwin
goarch: arm64
pkg: github.com/authzed/spicedb/internal/dispatch/singleflight
                       │ bits-and-blooms.txt │           boomfilters.txt           │
                       │       sec/op        │   sec/op     vs base                │
AddTestSerialize-12              4.415µ ± 0%   5.544µ ± 0%  +25.59% (p=0.000 n=10)
AddTestSerializeGob-12           4.407µ ± 0%   5.562µ ± 0%  +26.20% (p=0.000 n=10)
geomean                          4.411µ        5.553µ       +25.89%

                       │ bits-and-blooms.txt │           boomfilters.txt            │
                       │        B/op         │     B/op      vs base                │
AddTestSerialize-12             1.406Ki ± 0%   1.586Ki ± 0%  +12.78% (p=0.000 n=10)
AddTestSerializeGob-12          1.406Ki ± 0%   1.586Ki ± 0%  +12.78% (p=0.000 n=10)
geomean                         1.406Ki        1.586Ki       +12.78%

                       │ bits-and-blooms.txt │            boomfilters.txt            │
                       │      allocs/op      │  allocs/op   vs base                  │
AddTestSerialize-12               10.00 ± 0%   123.00 ± 0%  +1130.00% (p=0.000 n=10)
AddTestSerializeGob-12            10.00 ± 0%   123.00 ± 0%  +1130.00% (p=0.000 n=10)
geomean                           10.00         123.0       +1130.00%
```

Reply:
For reference, the tests:

# bits-and-blooms

```go
package singleflight

import (
	"github.com/bits-and-blooms/bloom/v3"
	"testing"
)

var s = []string{
	"3#JQ9s7@",
	"f^&4L6$K",
	"2%hN5*G@",
	"p$#8q9!W",
	"1@L6t#E$",
	"m^&7P5!S",
	"4%kN3*H@",
	"j$#2r9@Q",
	"5@F6y#D$",
	"n^&1T7!U",
	"6%gM4*J@",
	"i$#3o9@R",
	"7@B5u#C$",
	"l^&2V6!X",
	"8%aK1*Z@",
	"h$#4m9@S",
	"9@G7x#Y$",
	"o^&5I3!N",
	"0%bL2*F@",
	"g$#6n9@T",
	"a@E8w#X$",
	"k^&4O1!M",
	"b%jP3*K@",
	"e$#7q9@U",
	"c@D5v#Z$",
	"d^&8R2!H",
	"f%tN4*L@",
	"i$#9s7@Q",
	"g@J6y#E$",
	"h^&5T3!U",
	"j%oM1*K@",
	"k$#8p9@R",
	"l@B7u#F$",
	"m^&2V5!X",
	"n%lK3*Z@",
	"o$#1m9@S",
	"p@G4x#C$",
	"q^&6I2!N",
	"r%bL4*F@",
	"s$#3n9@T",
	"t@E5w#Y$",
	"u^&7O1!M",
	"v%jP2*K@",
	"w$#4q9@U",
	"x@D6v#Z$",
	"y^&9R3!H",
	"z%tN5*L@",
}

func BenchmarkAddTestSerialize(b *testing.B) {
	for i := 0; i < b.N; i++ {
		bloomFilter := bloom.NewWithEstimates(defaultBloomFilterSize, 0.001)
		for _, randStr := range s {
			bloomFilter = bloomFilter.AddString(randStr)
		}
		for _, randStr := range s {
			bloomFilter.TestString(randStr)
		}
		_, _ = bloomFilter.MarshalBinary()
	}
}

func BenchmarkAddTestSerializeGob(b *testing.B) {
	for i := 0; i < b.N; i++ {
		bloomFilter := bloom.NewWithEstimates(defaultBloomFilterSize, 0.001)
		for _, randStr := range s {
			bloomFilter = bloomFilter.AddString(randStr)
		}
		for _, randStr := range s {
			bloomFilter.TestString(randStr)
		}
		_, _ = bloomFilter.GobEncode()
	}
}
```

# boomfilters

```go
package singleflight

import (
	"bytes"
	boom "github.com/tylertreat/BoomFilters"
	"testing"
)

var s = []string{
	"3#JQ9s7@",
	"f^&4L6$K",
	"2%hN5*G@",
	"p$#8q9!W",
	"1@L6t#E$",
	"m^&7P5!S",
	"4%kN3*H@",
	"j$#2r9@Q",
	"5@F6y#D$",
	"n^&1T7!U",
	"6%gM4*J@",
	"i$#3o9@R",
	"7@B5u#C$",
	"l^&2V6!X",
	"8%aK1*Z@",
	"h$#4m9@S",
	"9@G7x#Y$",
	"o^&5I3!N",
	"0%bL2*F@",
	"g$#6n9@T",
	"a@E8w#X$",
	"k^&4O1!M",
	"b%jP3*K@",
	"e$#7q9@U",
	"c@D5v#Z$",
	"d^&8R2!H",
	"f%tN4*L@",
	"i$#9s7@Q",
	"g@J6y#E$",
	"h^&5T3!U",
	"j%oM1*K@",
	"k$#8p9@R",
	"l@B7u#F$",
	"m^&2V5!X",
	"n%lK3*Z@",
	"o$#1m9@S",
	"p@G4x#C$",
	"q^&6I2!N",
	"r%bL4*F@",
	"s$#3n9@T",
	"t@E5w#Y$",
	"u^&7O1!M",
	"v%jP2*K@",
	"w$#4q9@U",
	"x@D6v#Z$",
	"y^&9R3!H",
	"z%tN5*L@",
}

func BenchmarkAddTestSerialize(b *testing.B) {
	for i := 0; i < b.N; i++ {
		var buf bytes.Buffer
		bf := boom.NewUnstableBloomFilter(defaultBloomFilterSize, 0.001)
		for _, randStr := range s {
			bf.Add([]byte(randStr))
		}
		for _, randStr := range s {
			bf.Test([]byte(randStr))
		}
		_, _ = bf.WriteTo(&buf)
	}
}

func BenchmarkAddTestSerializeGob(b *testing.B) {
	for i := 0; i < b.N; i++ {
		bf := boom.NewUnstableBloomFilter(defaultBloomFilterSize, 0.001)
		for _, randStr := range s {
			bf.Add([]byte(randStr))
		}
		for _, randStr := range s {
			bf.Test([]byte(randStr))
		}
		_, _ = bf.GobEncode()
	}
}
```